ImageSampler : Add GafferScene ImageSampler node

ericmehl opened the pull request:
This pull request adds an ImageSampler node to sample image data onto geometry primitive variables.
- Samples one or more channels from a GafferImage onto a single primitive variable.
- Automatically creates appropriate type and interpretation for Cs, N, P, Pref, scale, uv, velocity and width primitive variables (this list based on the preset hints in the OSLObject output creator).
- Non-automatic variables are suffixed with a "." for each image channel sampled.

### Related issues ###

This is a feature addition based on Gaffer [forum thread](https://groups.google.com/g/gaffer-dev/c/8k_3Yk3UtH4/m/g62FoLosBwAJ).

### Dependencies ###

None

### Breaking changes ###

None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/master/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
Apologies for the double PR - this is a fixup of an [earlier PR](https://github.com/GafferHQ/gaffer/pull/3826) with some relevant comments from @johnhaddon.

In this update, I fixed the DocumentationTest and I changed the compute to single threaded. It sounds like multi-threading Sampler is a little out of scope here. I haven't done extensive profiling but in production I've found that the bottleneck in image sampling is with the Sampler / reading the image data even for relatively large models (on the order of 3 million vertices).

johnhaddon commented on src/GafferScene/ImageSampler.cpp:
```diff
@@ -0,0 +1,426 @@
+	{
+		for( size_t i = 0; i < channels.size(); ++i )
+		{
+			if( auto o = addPrimitiveVariable( outputPrimitive.get(), primVarName + "." + channels[i], interpolation ) )
```
I found this a bit confusing, particularly in the PrimitiveInspector where the multiple `prefix.suffix` primvars appear alongside single vector primvars which are labelled as `prefix.suffix` on a per-column basis. Is it especially useful to be able to sample multiple float channels when we can't sample multiple V3f/V2d channels etc? Is there merit in a simpler "one node, one primvar" rule?

ericmehl commented:
No worries on timing, that's great news! Thanks for making the review, it's really helpful and a lot of good improvements to make.

I think you are right about being able to sample to an arbitrary vector, color, etc. so that's definitely a change I'll make. I originally was limiting it to sampling a single variable because I was struggling to think of an interface that wouldn't get overly complicated with associating channels to variables, vector indices and such.

Now though I think there might be a very simple solution along the lines of the CustomAttributes node. What do you think about mimicking the UI for adding a custom attribute, but instead of having for example, float plugs, it would be string plugs where the user can input the channel name? I'd pass empty vector indices unchanged if there is no channel name for that index, or default to zero if there is no existing primitive variable.

That would also solve the confusing prefix.suffix in https://github.com/GafferHQ/gaffer/pull/3827#discussion_r458022527

This would solve a use-case I've been handling through additional primitive variable shuffle nodes downstream for sampling one channel onto a single axis of "P". The motivation for this node has been sampling heightfield data and seed densities onto a grid for dynamic landscape generation. The scheme above would allow creating a V3f "P" variable, adding a channel to P.y and leaving x and z empty to leave their values unchanged.

How does that sound? I'm definitely interested to hear from @murrays-ie if you have time.

ericmehl commented:
One other question - I'm considering trying an optimization where I presort the points based on their uv values into buckets that correspond to image tiles and then processing those tiles in order. In my Gaffer script where I'm sampling heightfields I find I've had to increase the hash cache limit quite a bit to avoid huge performance bottlenecks when sampling images. 

I don't know for sure, but I'm speculating that when I use the ImageSampler to sample randomized seeds from large images it is getting so many image tiles that it fills the cache and causes it to have to frequently re-hash tiles.

Maybe if I find it gives a decent performance increase it's worth adding as a toggle? I'd expect many use cases would have a somewhat built-in binning of uv just based on the mesh topology, but cases like the scattered points would not.

I'm guessing I would need to try it and see how performance goes, but I thought I'd see if you or others have an opinion on that route.

ericmehl commented on src/GafferScene/ImageSampler.cpp:
```diff
@@ -0,0 +1,426 @@
+			o(  i, 
+				channelIndex, 
+				sampler,
+				( uv.x * ( ( float )imageSize.x - 1.0f ) + imageMin.x ) + 0.5f, 
+				( uv.y * ( ( float )imageSize.y - 1.0f ) + imageMin.y ) + 0.5f
```
The `Sampler` documentation bit here is where I get the idea for adding the `0.5` offset: https://github.com/GafferHQ/gaffer/blob/fd073b7e05393519fd87adb03ecb230b4bb62646/include/GafferImage/Sampler.h#L85-L96

I think as a user I'd expect a 4x4 pixel image sampled onto a 4x4 point grid to sample pixels exactly, so the `0.5` offset is needed, right? Or maybe my brain is inside out?

johnhaddon commented:
> Now though I think there might be a very simple solution along the lines of the CustomAttributes node. What do you think about mimicking the UI for adding a custom attribute, but instead of having for example, float plugs, it would be string plugs where the user can input the channel name?

That sounds pretty good to me, if you're up for doing the additional work.

> I'd pass empty vector indices unchanged if there is no channel name for that index, or default to zero if there is no existing primitive variable.

This I'm less sure about, and it makes me wonder if we're imagining exactly the same thing. I'm imagining that each row would contain `name, type, channels` as follows :

`name` : A StringPlug specifying the name of the primitive variable to create. We would always overwrite this primitive variable entirely. It wouldn't be permissible to enter `P.y` here.
`type` : A dropdown with `Color/Vector/Point/Normal/Float`.
`channels` : A StringPlug where you enter a space-separated list of channels. The number of channels you need to enter is dictated by the `type`.

We could then customise the `+` menu button so that it had convenient items for `Cs`, `P` etc in the same way as we do for OSLObject. So we'd have the convenience of your original design, but with the flexibility of arbitrary numbers and types of variables.

> This would solve a use-case I've been handling through additional primitive variable shuffle nodes downstream for sampling one channel onto a single axis of "P".

When writing new nodes to facilitate a particular workflow, I think there's a danger that you tailor the node too much to that initial use case, rather than design it to interoperate well with other nodes for multiple use cases. I may be wrong here, and feel free to make a counterargument, but I think this "write a subset of components" thing is an example. Perhaps in this case it's useful to overwrite P.y directly, but in others you might need to apply an offset or any number of weird and wonderful things. I think this is why it's appropriate to keep ImageSampler in line with PrimitiveSampler/CopyPrimitiveVariables/ShufflePrimitiveVariables etc, by operating simply on whole primitive variables, with no interleaving. Then we defer to OSLObject to provide the full generality of per-component operations, merging and math etc.

> One other question - I'm considering trying an optimization where I presort the points based on their uv values into buckets that correspond to image tiles and then processing those tiles in order. In my Gaffer script where I'm sampling heightfields I find I've had to increase the hash cache limit quite a bit to avoid huge performance bottlenecks when sampling images.

That would certainly be an interesting experiment. A related idea would be to not process all the tiles for one channel before moving on to another channel. The ColorProcessor nodes need to calculate corresponding RGB channels together, and the ImageReader likes to load multiple channels for a single tile at once too. So you might get much better cache coherency by constructing the samplers for all the channels upfront and then having the channel loop be the innermost one. I'd be very interested in seeing the kind of network you're using to process the heightfields, and the resolution of the images involved to see if there's anything we can do in the core too.

> Maybe if I find it gives a decent performance increase it's worth adding as a toggle? I'd expect many use cases would have a somewhat built-in binning of uv just based on the mesh topology, but cases like the scattered points would not.

My experience with "toggle this to go faster in certain circumstances" buttons has been that they often don't get used at the appropriate time. We don't always succeed, but our goal is simple controls and good performance rather than a plethora of things to tune. I suggest we get your node merged first, then do the testing with sorted points, and then make a call separately. The channels-in-the-inner-loop thing might be worth trying now though, since it's simple and shouldn't require a user toggle.

> Thanks for making the review, it's really helpful and a lot of good improvements to make.

No worries - thanks for being so receptive.

ericmehl commented:
> I'd be very interested in seeing the kind of network you're using to process the heightfields, and the resolution of the images involved to see if there's anything we can do in the core too.

I'm happy to share the script with you - there's a lot more data associated with it than just the heightfields though so it's not super straight forward. But I need to create an Amazon Web Services image with a ready-to-go version of it for my client though so I may be able to give you access through that. 

I didn't start running into problems until I added the vegetation instancing on top of the heightfield so it's not as easy as sending you the heightfield-only part. Would using Encapsulate on the instances reduce pressure on the hash cache? There are probaby about 100 unique scene locations and about 27 million instances in the end. As those numbers got higher I started seeing the ImageReader hash as the most time consuming plug in the performance monitor and worked around it by refactoring the graph. I certainly have more to learn about optimizing graphs in general.



> That sounds pretty good to me, if you're up for doing the additional work.

It seems like a big enough improvement to be worth it, so I'll take it on.

> `channels` : A StringPlug where you enter a space-separated list of channels. The number of channels you need to enter is dictated by the `type`.

The only difference I have in mind is instead of a single string plug, I'd add the number of string plugs corresponding to the size of the data type. Then you could optionally leave them blank and that would either pass through existing data or fill it with a default (presumably 0). I understand what you mean to keep it in the same style as the other sampler nodes and I don't have a particularly strong feeling about my imagined method.

> My experience with "toggle this to go faster in certain circumstances" buttons has been that they often don't get used at the appropriate time.

Agreed here too. I'll have a look at the sorting method and see how it affects performance. I'm guessing there won't be too big of a hit to bin them, especially since it should be able to be multi-threaded.

ericmehl commented:
Hi John,
I updated my ImageSampler node based on what we talked about a little while ago. It can sample multiple primitive variables in one node now and uses GeometricData.Interpretation to verify the number of channels. And I added the preset variables like OSLObject.

One [test is currently failing](https://github.com/GafferHQ/gaffer/blob/cd483e9d523a1ccf2230ac7e57c6a4a9b7776f86/python/GafferSceneTest/ImageSamplerTest.py#L390) and I could use a hand sorting it out. When I [store the Sampler](https://github.com/GafferHQ/gaffer/blob/cd483e9d523a1ccf2230ac7e57c6a4a9b7776f86/src/GafferScene/ImageSampler.cpp#L397-L398) ahead of time for each channel and then use that in the [inner sampler loop](https://github.com/GafferHQ/gaffer/blob/cd483e9d523a1ccf2230ac7e57c6a4a9b7776f86/src/GafferScene/ImageSampler.cpp#L464-L478), I'm losing context variables upstream. 

I don't think I'm changing context between the points when I create and use the Sampler, but there's definitely something not right.

Overall I think it's a solid improvement over sampling a single variable, I'm happy for your recommendation.

ericmehl commented:
I did a quick rebase to keep this PR from getting too far behind master, not meaning to put any pressure on reviewing it.

The context bug is still present from https://github.com/GafferHQ/gaffer/pull/3827#issuecomment-676630957 that I could use a bit of help sorting out.
